content/posts/2019-08-03---Going-from-ember-to-vue.md

```diff
@@ -0,0 +1,284 @@
+
+Much like React, Vue is a component library -- which means you'll be using components for <strong>everything.</strong>
+That means no controllers or route objects.
+When you create a route, you are pointing to a component.
+There is literally nothing special about this component -- it can be literally any component that you want (in fact, you can use it both as a route _and_ component on the page... although, you really shouldn't).
+
+In my opinion, this is a good habit to get into as it will help you transition to something like React down the line if you have to.
+
+## Batteries aren't included
+
+![Batteries](https://i.imgur.com/65SKMnM.jpg?1)
+<aside>Photo by <a href="https://www.pexels.com/@hilaryh">Hilary Halliwell on Pexels</a></aside>
+
+This isn't meant to be a knock on Vue, but it doesn't come with as much stuff out of the box as Ember.
+In fact, this is probably a huge plus.
+[Vue CLI](https://cli.vuejs.org/) allows you to generate a project quickly, but much like Create React App, you are getting a pretty simple UI application with webpack pre-configured for you.
+And yes, webpack over Brocolli is another big difference... but it's probably a good one.
+
+However, anything you want to do in ember, probably already has an equivalent library in Vue.
+And most of these libraries are maintained by core team members, so they feel very official.
+With just a couple more `vue add` commands, you can get routing and state management in your app, so it's really not a _huge_ knock against Vue.
+
+## Components are one file
+
+I found this weird at first, but boy, do I like components with both markup and JavaScript in the same file.
+I hated having to bounce between files so often in Ember (especially when the community wanted to move away from pod structure).
+
+<figure>
+  <blockquote>
+    <p>Ugh, markup and JS in the same file?? I believe in separation of concerns.</p>
+    <footer>
+      <cite>— People who want you to know they're smart.</cite>
+    </footer>
+  </blockquote>
+</figure>
+
+I've actually heard this complaint a few times about "I don't like JS/HTML/CSS in the same file because I like separation of concerns."
+I get the sentiment, but to be clear, just because "concerns" are in the same file, doesn't mean they aren't separated.
+The JavaScript of a component is too tightly tied to the markup (as it should be), so why force yourself to bounce between the two files?
+
+## State management
+
+The popular state management solution seems to be [Vuex](https://vuex.vuejs.org/) (another "official" Vue library written by core team members).
+This is much closer to [Redux](https://redux.js.org/) or [NGRX](https://ngrx.io/).
+I'm not going to claim that these patterns are the end-all-be-all solutions for state management on the front-end, but it's where most of the community seems to be going right now.
+Which is a huge plus over Ember.
+The skills/practices you pick up writing Vuex will translate to these other libraries.
```

Comment:
Is there an idea of building "relationships" into the identity map?  Also, is there a plug and play with different standard response formats like JSONAPI or REST?

Reply:
No, this would be something you'd manage yourself or use a package (for both serializing data and building relationships).  Relating models has felt more like more of a headache but doing a-typical things like member-actions or supporting non JSONAPI payloads has felt a lot easier.

Reply:
it's pretty easy in ember to do off-handing things, too. Just usin' `fetch` and async/await :)

Reply:
For consuming JSONAPI you can use [spraypaint](https://github.com/graphiti-api/spraypaint-vue). It's actually pretty agnostic so you can use it with (P)React or other framework as well.

There are also vuex related projects for more complex apps like [vuex orm](https://vuex-orm.github.io/vuex-orm/) with its [axios](https://github.com/vuex-orm/plugin-axios) or [graphql](https://github.com/vuex-orm/plugin-graphql) plugins. It's more like Ember data for Vuex :).

Reply:
Much like a lot of the things in this post, the answer is "yeah, you can find a way to do it because it's JavaScript."  I'm sure there are libraries for every framework that help to support JSONAPI but I want to look at what comes officially supported by the Vue team.

Reply:
Probably you won't find it as Evan once said: 

> Vue doesn't offer opinionated data layer, because in my opinion such heavy-handed data modeling is unnecessary for the client. The preferred client-side state management solution is Vuex. But if you like Ember data you can use something like http://www.js-data.io/ with Vue."
> 
> source: https://www.reddit.com/r/vuejs/comments/53ges7/coming_from_ember_to_vue_what_to_expect/d7td2sy